### PR TITLE
Add cache for multipart uploads of an object

### PR DESCRIPTION
**Pull request by shtripat**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
This cache would retain the parts in cache for 24hrs and then evict them.

## Motivation and Context
Maintain a cache of uploaded parts of a objects for listing them overall, by `prefix` and by `upalodIDMarker` etc

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/erasure-server-pool.go [hunk 1]
```diff
@@ -47,6 +47,72 @@ import (
...
+}
+
+// cleanExpiredEntries removes expired entries every hour
+func (c *MultipartUploadCache) cleanExpiredEntries() {
+	for {
+		time.Sleep(1 * time.Hour)
+		c.mutex.Lock()
+		now := time.Now()
+		for key, entry := range c.data {
+			if now.Sub(entry.Timestamp) > 24*time.Hour {
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]:
```suggestion
			if now.Sub(entry.Timestamp) > globalAPIConfig.getStaleUploadsExpiry() {
```

File: cmd/erasure-server-pool.go [hunk 2]
```diff
@@ -47,6 +47,72 @@ import (
...
+		Timestamp: time.Now(),
+	}
+}
+
+// Get retrieves a value from the cache if it has not expired
+func (c *MultipartUploadCache) Get(key string) (MultipartUpload, bool) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	entry, exists := c.data[key]
+	if !exists || time.Since(entry.Timestamp) > 24*time.Hour {
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]:
```suggestion
	if !exists || time.Since(entry.Timestamp) > globalAPIConfig.getStaleUploadsExpiry() {
```

File: cmd/erasure-server-pool.go [hunk 3]
```diff
@@ -47,6 +47,72 @@ import (
...
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.data[key] = MultipartUploadCacheEntry{
+		Upload:    value,
+		Timestamp: time.Now(),
+	}
+}
+
+// Get retrieves a value from the cache if it has not expired
+func (c *MultipartUploadCache) Get(key string) (MultipartUpload, bool) {
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]:
```suggestion
func (c *MultipartUploadCache) Get(key string) (MultipartUpload, bool) {
```

Add another helper called `Has(key string) (bool) {`

File: cmd/erasure-server-pool.go [hunk 4]
```diff
@@ -1713,8 +1785,16 @@ func (z *erasureServerPools) ListMultipartUploads(ctx context.Context, bucket, p
 		if err != nil {
 			return result, err
 		}
-		poolResult.Uploads = append(poolResult.Uploads, result.Uploads...)
+		// poolResult.Uploads = append(poolResult.Uploads, result.Uploads...)
+		uploads = append(uploads, result.Uploads...)
+	}
+
+	for _, upload := range uploads {
+		if _, found := z.multipartUploadsCache.Get(upload.UploadID); found {
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]:
```suggestion
		if ok := z.multipartUploadsCache.Has(upload.UploadID); ok {
```

implement a Has(key) method to make this simpler

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/erasure-server-pool.go [hunk 5]
```diff
@@ -47,6 +47,80 @@ import (
...
+		Timestamp: time.Now(),
+	}
+}
+
+// Get retrieves a value from the cache if it has not expired
+func (c *MultipartUploadCache) Get(key string) (MultipartUpload, bool) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	entry, exists := c.data[key]
+	if !exists || time.Since(entry.Timestamp) > globalAPIConfig.getStaleUploadsExpiry() {
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]: Get expiry value before the lock.

File: cmd/erasure-server-pool.go [hunk 6]
```diff
@@ -47,6 +47,80 @@ import (
...
+}
+
+// cleanExpiredEntries removes expired entries every hour
+func (c *MultipartUploadCache) cleanExpiredEntries() {
+	for {
+		time.Sleep(1 * time.Hour)
+		c.mutex.Lock()
+		now := time.Now()
+		for key, entry := range c.data {
+			if now.Sub(entry.Timestamp) > globalAPIConfig.getStaleUploadsExpiry() {
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]: Get expiry before the lock?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/erasure-server-pool.go (hunk 5 above)
Comment (shtripat) [reply cmd/erasure-server-pool.go]: I thought like get entry from map after getting a lock. Do you suggest like
```
        entry, exists := c.data[key]
        c.mutex.RLock()
        defer c.mutex.RUnlock()
        if !exists || time.Since(entry.Timestamp) > globalAPIConfig.getStaleUploadsExpiry() {
                return MultipartUpload{}, false
        }
        return entry.Upload, true
```
not sure I understood.

Comment (harshavardhana) [inline cmd/erasure-server-pool.go]:
```suggestion
        expiry := globalAPIConfig.getStaleUploadsExpiry()
	c.mutex.RLock()
	defer c.mutex.RUnlock()
	entry, exists := c.data[key]
	if !exists || time.Since(entry.Timestamp) > expiry {
```

**Review by harshavardhana — COMMENTED** [review]
this is what I meant

File: cmd/erasure-server-pool.go (hunk 5 above)
Comment (shtripat) [reply cmd/erasure-server-pool.go]: ah ok, sorry for being so naive. Thanks.

File: cmd/erasure-server-pool.go [hunk 7]
```diff
@@ -1791,6 +1881,13 @@ func (z *erasureServerPools) PutObjectPart(ctx context.Context, bucket, object,
 		return PartInfo{}, err
 	}
 
+	// Add the object part to the cache
+	z.multipartUploadsCache.Set(uploadID, MultipartUpload{
+		Bucket:   bucket,
+		Object:   object,
+		UploadID: uploadID,
+	})
+
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]: the multipart uploadID must be set in newMultipartUpload() and 
removed from the map when 

- CompleteMultipart happens
- or AbortMultipart happens

We also need a peer call where we can notify our peers that they need to delete their local uploadID cache in memory.

Comment (harshavardhana) [reply cmd/erasure-server-pool.go]: It should never be set for every PutObjectPart()

File: cmd/erasure-server-pool.go [hunk 8]
```diff
@@ -47,6 +47,82 @@ import (
...
+		c.mutex.Lock()
+		now := time.Now()
+		for key, entry := range c.data {
+			if now.Sub(entry.Timestamp) > expiry {
+				delete(c.data, key)
+			}
+		}
+		c.mutex.Unlock()
+	}
+}
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]: I think what we can do is rely on cleanupStaleUploadsOnDisk() to return []string of uploadIDs, this uploadIDs you can work on the delete from the map instead of doing a new goroutine here. 

```
func (er erasureObjects) cleanupStaleUploadsOnDisk(ctx context.Context, disk StorageAPI) {
```

Comment (harshavardhana) [reply cmd/erasure-server-pool.go]:
```go
func (s *erasureSets) cleanupStaleUploads(ctx context.Context) {
        timer := time.NewTimer(globalAPIConfig.getStaleUploadsCleanupInterval())
        defer timer.Stop()

        for {
                select {
                case <-ctx.Done():
                        return
                case <-timer.C:
                        var wg sync.WaitGroup
                        for _, set := range s.sets {
                                wg.Add(1)
                                go func(set *erasureObjects) {
                                        defer wg.Done()
                                        if set == nil {
                                                return
                                        }
                                        set.cleanupStaleUploads(ctx)
                                }(set)
                        }
                        wg.Wait()

                        // Reset for the next interval
                        timer.Reset(globalAPIConfig.getStaleUploadsCleanupInterval())
                }
        }
}
```

File: cmd/erasure-server-pool.go (hunk 7 above)
Comment (shtripat) [reply cmd/erasure-server-pool.go]: oh ok, will change accordingly.

File: cmd/erasure-server-pool.go [hunk 9]
```diff
@@ -47,6 +47,87 @@ import (
...
+
+// Remove removes given entry from cache
+func (c *MultipartUploadCache) Remove(key string) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	delete(c.data, key)
+}
+
+// cleanExpiredEntries removes expired entries every hour
+func (c *MultipartUploadCache) cleanExpiredEntries(z *erasureServerPools) {
```
Comment (harshavardhana) [inline cmd/erasure-server-pool.go]: You don't need this anymore

**Review by harshavardhana — CHANGES_REQUESTED** [review]
You would also need to modify to notify peers.

File: cmd/erasure-server-pool.go (hunk 9 above)
Comment (shtripat) [reply cmd/erasure-server-pool.go]: Is it because, while Abort and Complete we are removing entries from cache?
Also not clear on usage of returned `[]string` of cleaned up upload ids from `cleanupStaleUploads`

**Comment by shtripat** [comment]
> You would also need to modify to notify peers.

Is it like `[]string` returned from `cleanupStaleUploads` to be sent across peers for their removal?

File: cmd/erasure-server-pool.go (hunk 9 above)
Comment (harshavardhana) [reply cmd/erasure-server-pool.go]: why not? if you have two then they will conflict, you only need one source of truth.

**Comment by harshavardhana** [comment]
> Is it like `[]string` returned from `cleanupStaleUploads` to be sent across peers for their removal?

Correct the ones you just expired, you are letting others know that you did that if they have any local reference to remove it as well.

**Comment by harshavardhana** [comment]
Remember newMultipart can happen on one server and complete multipart can happen on another server and abort also can happen on another server, so you would end up with stale UploadIDs on different nodes.

**Comment by shtripat** [comment]
> > Is it like `[]string` returned from `cleanupStaleUploads` to be sent across peers for their removal?
> 
> Correct the ones you just expired, you are letting others know that you did that if they have any local reference to remove it as well.

The last commit, I am sending a notification to other nodes for their cleanup.

File: cmd/erasure-server-pool.go [hunk 10]
```diff
@@ -47,6 +47,71 @@ import (
...
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.data[key] = MultipartUploadCacheEntry{
+		Upload:    value,
+		Timestamp: time.Now(),
+	}
+}
+
+// Get retrieves a value from the cache if it has not expired
+func (c *MultipartUploadCache) Get(key string) (MultipartUpload, bool) {
```
Comment (klauspost) [inline cmd/erasure-server-pool.go]: Is this called from anywhere?

File: cmd/erasure-server-pool.go [hunk 11]
```diff
@@ -1713,8 +1784,16 @@ func (z *erasureServerPools) ListMultipartUploads(ctx context.Context, bucket, p
...
-		poolResult.Uploads = append(poolResult.Uploads, result.Uploads...)
+		// poolResult.Uploads = append(poolResult.Uploads, result.Uploads...)
+		uploads = append(uploads, result.Uploads...)
 	}
+
+	for _, upload := range uploads {
+		if ok := z.multipartUploadsCache.Has(upload.UploadID); ok {
+			poolResult.Uploads = append(poolResult.Uploads, upload)
+		}
+	}
```
Comment (klauspost) [inline cmd/erasure-server-pool.go]: I don't understand how this works.

A) The `multipartUploadsCache` is local only, correct? I don't see anything distributing cached entries or any hashing to have it placed dterministically.
B) An upload will only be cached on one server in cluster. How can we filter, if the upload will only be cached on a random server?
C) What if a server was restarted?

I may be missing something, but this seems to assume that the NewMultipartUpload and ListMultipartUploads happen on the same server, which is unlikely.

Both distributing to all servers and asking all servers seems like a non-scalable approach.

**Review by klauspost — CHANGES_REQUESTED** [review]
I feel like I am missing something. Where does the caching come in?

I only see extra code being added for housekeeping (setting/deleting entries), but nothing that provides a speedup by using the cache.

**Comment by harshavardhana** [comment]
> I feel like I am missing something. Where does the caching come in?
> 
> I only see extra code being added for housekeeping (setting/deleting entries), but nothing that provides a speedup by using the cache.

This is not a speed up, it is to provide listMultipartUploads() output when there is no object name provided.

@klauspost

**Comment by harshavardhana** [comment]
This is mainly for fix the behavior of this API for certain section of applications like alluxio underfs.

**Comment by klauspost** [comment]
> This is mainly for fix the behavior of this API for certain section of applications like alluxio underfs.

Okie, but how does it work across a cluster?

**Comment by harshavardhana** [comment]
> Okie, but how does it work across a cluster?

it is just per node basis @klauspost we just have to send some non-empty list here, the values themselves don't need to be "cluster" consistent.

File: cmd/erasure-server-pool.go (hunk 10 above)
Comment (shtripat) [reply cmd/erasure-server-pool.go]: Not used, can be removed

**Comment by klauspost** [comment]
Conflicts and 

```
2024-08-09T10:16:09.0411135Z === RUN   TestListMultipartUploadsHandler
2024-08-09T10:16:09.0412005Z     bucket-handlers_test.go:490: Test 2: ErasureSD: Expected the response status to be `404`, but instead found `200`
2024-08-09T10:16:09.0413146Z     bucket-handlers_test.go:507: Test 2: ErasureSD: Expected the response status to be `404`, but instead found `200`
2024-08-09T10:16:09.0414256Z     bucket-handlers_test.go:490: Test 2: Erasure: Expected the response status to be `404`, but instead found `200`
2024-08-09T10:16:09.0415357Z     bucket-handlers_test.go:507: Test 2: Erasure: Expected the response status to be `404`, but instead found `200`
2024-08-09T10:16:09.0416171Z --- FAIL: TestListMultipartUploadsHandler (0.14s)
```

**Comment by shtripat** [comment]
Looks like mint is not happy. will debug and send separate PR for the same.

File: cmd/bucket-handlers_test.go [hunk 12]
```diff
@@ -369,7 +369,7 @@ func testListMultipartUploadsHandler(obj ObjectLayer, instanceType, bucketName s
 			maxUploads:         "0",
 			accessKey:          credentials.AccessKey,
 			secretKey:          credentials.SecretKey,
-			expectedRespStatus: http.StatusNotFound,
```
Comment (harshavardhana) [inline cmd/bucket-handlers_test.go]: why is this changed? a non-existent bucket must return 404

**Review by harshavardhana — COMMENTED** [review]
A non-existent bucket must return error for ListMultipartUploads()

File: cmd/bucket-handlers_test.go (hunk 12 above)
Comment (shtripat) [reply cmd/bucket-handlers_test.go]: Test were failing as minio returning 200 OK in this case. I feel it's because we return from cache now.

**Comment by shtripat** [comment]
> A non-existent bucket must return error for ListMultipartUploads()

Oh ok, so need to add a check for bucket existence and return 404 if bucket does not exist, right?

File: cmd/peer-rest-server.go [hunk 13]
```diff
@@ -905,6 +906,23 @@ func (s *peerRESTServer) ReloadPoolMetaHandler(mss *grid.MSS) (np grid.NoPayload
...
+	if objAPI == nil {
+		return np, grid.NewRemoteErr(errServerNotInitialized)
+	}
+
+	pools, ok := objAPI.(*erasureServerPools)
+	if !ok {
+		return np, grid.NewRemoteErr(errors.New("not a pooled setup"))
+	}
+
+	uploadID := mss.Get(peerRESTCleanUploadID)
```
Comment (vadmeste) [inline cmd/peer-rest-server.go:920]: can we return an error if peerRESTCleanUploID is empty ? (easier for debugging issues)

File: cmd/erasure-server-pool.go [hunk 14]
```diff
@@ -47,6 +47,59 @@ import (
...
+func (c *MultipartUploadCache) Has(key string) bool {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	_, exists := c.data[key]
+	return exists
+}
+
+// Remove removes given entry from cache
+func (c *MultipartUploadCache) Remove(key string) {
+	c.mutex.RLock()
```
Comment (vadmeste) [inline cmd/erasure-server-pool.go]: s/RLock/Lock

File: cmd/erasure-server-pool.go [hunk 15]
```diff
@@ -47,6 +47,59 @@ import (
...
+	defer c.mutex.RUnlock()
```
Comment (vadmeste) [inline cmd/erasure-server-pool.go]: s/RLock/Lock

File: cmd/erasure-server-pool.go [hunk 16]
```diff
@@ -1706,15 +1764,39 @@ func (z *erasureServerPools) ListMultipartUploads(ctx context.Context, bucket, p
...
-
 	poolResult := ListMultipartsInfo{}
 	poolResult.MaxUploads = maxUploads
 	poolResult.KeyMarker = keyMarker
 	poolResult.Prefix = prefix
 	poolResult.Delimiter = delimiter
+
+	// if no prefix provided, return the list from cache
+	if prefix == "" {
+		for _, entry := range z.multipartUploadsCache.data {
```
Comment (vadmeste) [inline cmd/erasure-server-pool.go:1773]: I guess access to data should be protected

**Review by vadmeste — COMMENTED** [review]
Can you elaborate more what this PR does ? 

Also it looks like we need to remember that this can be a round robin loaded balancer cluster, if I am not mistaken, ListMultipartUploads will return empty when the request lands to a random node.

**Comment by shtripat** [comment]
> Can you elaborate more what this PR does ?
> 
> Also it looks like we need to remember that this can be a round robin loaded balancer cluster, if I am not mistaken, ListMultipartUploads will return empty when the request lands to a random node.

https://github.com/minio/minio/pull/20181#issuecomment-2269565179 explains it. Its node local cache and mainly for fix the behavior of this API for certain section of applications like alluxio underfs.

File: cmd/peer-rest-server.go (hunk 13 above)
Comment (shtripat) [reply cmd/peer-rest-server.go:920]: Ack. Will add the same

File: cmd/erasure-server-pool.go (hunk 16 above)
Comment (shtripat) [reply cmd/erasure-server-pool.go:1773]: Ack

**Comment by harshavardhana** [comment]
> Also it looks like we need to remember that this can be a round robin loaded balancer cluster, if I am not mistaken, ListMultipartUploads will return empty when the request lands to a random node.

the idea is not meant to be a consistent view of everything that is present at the multipart uploads namespace. IT is mainly to satisfy a class of applications such as alluxio/underfs. 

just like how we have done changes for spark/hadoop specific behavior.

**Comment by harshavardhana** [comment]
there are bunch of issues with this PR addressed them and simplified the approach here https://github.com/minio/minio/pull/20407

**Comment by shtripat** [comment]
Closing in favor of https://github.com/minio/minio/pull/20407
